xchacha20poly1035/xchachapoly.go

```diff
@@ -0,0 +1,237 @@
+package xchacha20poly1305
+
+import (
+	"crypto/cipher"
+	"encoding/binary"
+	"errors"
+
+	"golang.org/x/crypto/chacha20poly1305"
+)
+
+var sigma = [4]uint32{0x61707865, 0x3320646e, 0x79622d32, 0x6b206574}
+
+type xchacha20poly1305 struct {
+	key [KeySize]byte
+}
+
+const (
+	// KeySize is the size of the key used by this AEAD, in bytes.
+	KeySize = 32
+	// NonceSize is the size of the nonce used with this AEAD, in bytes.
+	NonceSize = 24
+)
+
+//New xChaChapoly1305 AEAD with 24 byte nonces
+func New(key []byte) (cipher.AEAD, error) {
+	if len(key) != KeySize {
+		return nil, errors.New("chacha20poly1305: bad key length")
+	}
+	ret := new(xchacha20poly1305)
+	copy(ret.key[:], key)
+	return ret, nil
+
+}
+func (c *xchacha20poly1305) NonceSize() int {
+	return NonceSize
+}
+
+func (c *xchacha20poly1305) Overhead() int {
+	return 16
+}
+
+func (c *xchacha20poly1305) Seal(dst, nonce, plaintext, additionalData []byte) []byte {
+	if len(nonce) != NonceSize {
+		panic("xchacha20poly1305: bad nonce length passed to Seal")
+	}
+
+	if uint64(len(plaintext)) > (1<<38)-64 {
+		panic("xchacha20poly1305: plaintext too large")
+	}
+
+	var subKey [KeySize]byte
+	var hNonce [16]byte
+	var subNonce [chacha20poly1305.NonceSize]byte
+	copy(hNonce[:], nonce[:16])
+
+	HChaCha20(&subKey, &hNonce, &c.key)
+
+	chacha20poly1305, _ := chacha20poly1305.New(subKey[:])
+
+	copy(subNonce[4:], nonce[16:])
+
+	return chacha20poly1305.Seal(dst, subNonce[:], plaintext, additionalData)
```

Comment:
so chacha only accepts a 12 byte nonce. but we have a 24 byte nonce.

so we take the first 16 bytes, and use it as a nonce with our encryption key through Hchacha20 to get a new 32-byte encryption key (`subKey`).

Then we can encrypt using our new key (`subKey`) and a 12 byte nonce containing the final 8 bytes from the original nonce.

Is that all right? 

And the reason this is good is because originally, we were limited to to 2^96 msgs with a given key due to the 12-byte nonce. but now, even though the nonce used in chacha20poly is still actually only 8 bytes, the other 16 bytes were used to alter the key, so we get a full 2^192 msgs per original key (ie 2^64 msgs per subkey, and 2^128 subkeys per original key).

Is that right ?

Reply:
As for hchacha20, is the only guarantee we need from it collision resistance ?

Reply:
yes this is correct